### ISSUE TITLE
[lit-next] help with parse5/dom5

### ISSUE THREAD

**Issue by 43081j** [lit/lit#1418]
regarding this TODO:

https://github.com/Polymer/lit-html/blob/4395e3d45d68cf704155b507b2e0607fc7759a0c/packages/lit-ssr/src/lib/util/parse5-utils.ts#L15-L21

I am currently trying to fix parse5's types at DefinitelyTyped/DefinitelyTyped#48402 but have so far really struggled to get any reviewers.

I have also been working on a dom5 update at polymer/tools#3537 which is blocked by the above PR.

since it seems you need the same changes i've been making, would it be possible any of you could help review these? i can then open a lit-next PR to pull the newly published dom5 in once its ready.

the dom5 branch currently builds and passes tests with the new types, but ci fails since it doesn't yet have those new types (i copied them over locally). but it needs good review as a fair amount changed that i don't have the full picture of (since i haven't used much of dom5).

similarly, is it worth moving dom5 into its own repo or the lit repo? since the rest of polymer/tools is pretty stale these days.

cc @usergenic

**Comment by sorvell** [comment]
@usergenic Can you please provide some guidance on this? Thanks.

**Comment by LarsDenBakker** [comment]
Getting good types for parse5 and dom5 would be great indeed. FWIW we created a smaller subset of what dom5, on top of the existing tree adapter utils that parse5 ships. We monkey patch the types here to keep typescript happy.

**Comment by 43081j** [comment]
im <- -> this close to having new parse5 types, dom5, lit-next, and a bunch of other related things 😂 

the DT PR is a bit of a monolith, the only thing i really need a second pair of eyes on is how i've done strongly typed tree adapters.

you can basically now create a tree adapter type with a map interface:

```ts
interface MyTypeMap extends TreeAdapterTypeMap {
  node: MyNode;
  document: MyDocument;
}

declare const myAdapter: TypedTreeAdapter<MyTypeMap>;

const doc = parse('blah', {treeAdapter: myAdapter}); // MyDocument, NOT Document
const doc = parse('blah'); // Document
```

it feels like hackery using an interface map like that, but given how absurdly dynamic/weakly-typed parse5 is at the source, i couldn't think of a better option.

if someone can tell me that isn't a silly idea, at least then i know the PR makes sense.

**Comment by 43081j** [comment]
@sorvell in the end i got the new types for parse5 merged, and went ahead and updated dom5 in Polymer/tools#3537. However, as you can see the older stuff in the polymer tools repo then conflicts and also needs updating (a pretty big job).

i think it'd make more sense if dom5 was moved into this repo or polymerlabs, as a standalone package. once you guys get through the lit-next work going on, maybe that could be an option? would be happy to see it through once i know where it should live.

**Comment by justinfagnani** [comment]
@43081j I think a new repo makes sense - Polymer/tools can be updated to use the currently published version.

I wouldn't mind revisiting some of the APIs too - they often seem hard to differentiate - but maybe that's a very separate thing. We can start in PolymerLabs, but we'll also have a new location for things, or since SSR uses parts of it, maybe it belongs in the lit-html monorepo.

**Comment by 43081j** [comment]
That sounds good to me. A revisit of the APIs would be good too, even if in a separate PR. Would be happy to do those changes if there's some agreement on it too. 

Since it is used (or was until the types got outdated) by third parties unrelated to lit, I'd probably put it in polymer labs. 

If you can make that, I'm happy to open my pr there instead

**Comment by justinfagnani** [comment]
Since this was opened you became a maintainer of parse5, right @43081j?

**Comment by 43081j** [comment]
@justinfagnani thats true, the new version recently got published afaik too.

i think whats likely to happen at some point (ill probably do it) is have an equivalent to dom5 inside parse5 itself (e.g. `@parse5/tools`). that way those helpers can be maintained alongside the source of p5.
